[libcxx] adds a size-based representation for `vector`'s unstable ABI

cjdb opened the pull request:
**tl;dr** We can significantly improve the runtime performance of `std::vector` by changing its representation from three pointers to one pointer and two integers. This document explains the details of this change, along with the justifications for making it. See the [RFC] for more information.

This commit changes `std::vector`'s representation in a similar manner to `__split_buffer` in #139632. We introduce a layout type that tracks implementation details that differ between the pointer-based vector and size-based vector representations. `vector` does not parameterise its layout type, unlike `__split_buffer`. `__split_buffer` is used by both `vector` and `deque`, and being able to take an ABI break on `vector` doesn't automatically mean that users can also tolerate a break on `deque`. The most convenient way to work around this problem is to parameterise the layout type. Users of `std::vector` should not depend on its layout, and libc++ has no internal reason to toggle it, so we keep the representation concrete.

To use this feature, you'll need to configure libc++ so that it's built with `-DLIBCXX_UNSTABLE_ABI=Yes`.

[RFC]: https://discourse.llvm.org/t/adding-a-size-based-vector-to-libc-s-unstable-abi/86306

cjdb committed:
[libcxx] adds a size-based representation for `vector`'s unstable ABI

**tl;dr** We can significantly improve the runtime performance of
`std::vector` by changing its representation from three pointers to one
pointer and two integers. This document explains the details of this
change, along with the justifications for making it. See the [RFC] for
more information.

This commit changes `std::vector`'s representation in a similar manner
to `__split_buffer` in #139632. We introduce a layout type that tracks
implementation details that differ between the pointer-based vector
and size-based vector representations. `vector` does not parameterise
its layout type, unlike `__split_buffer`. `__split_buffer` is used by
both `vector` and `deque`, and being able to take an ABI break on
`vector` doesn't automatically mean that users can also tolerate a break
on `deque`. The most convenient way to work around this problem is to
parameterise the layout type. Users of `std::vector` should not depend
on its layout, and libc++ has no internal reason to toggle it, so we
keep the representation concrete.

To use this feature, you'll need to configure libc++ so that it's built
with `-DLIBCXX_UNSTABLE_ABI=Yes`.

[RFC]: https://discourse.llvm.org/t/adding-a-size-based-vector-to-libc-s-unstable-abi/86306

cjdb commented:
**This PR contains #139632 for CI purposes**. Only two files require review: `libcxx/include/__vector/vector.h` and `libcxx/test/std/algorithms/alg.modifying.operations/alg.swap/ranges.swap_ranges.pass.cpp`; the rest should evaporate once #139632 is merged with main.

The LLDB and GDB pretty-printers will require some tweaking again, but those changes haven't been applied yet.

cjdb commented:
@ldionne Would you mind reviewing this, please?

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -116,6 +426,12 @@ class vector {
+  using __base::back;
+  using __base::capacity;
+  using __base::empty;
+  using __base::size;
```
Although users aren't allowed to take the address of member functions from standard library types, there seems to be an endless supply of build breaks when I do this. The problem is that `&std::vector<T>::size` no longer exists, since the function is a member of `__vector_layout`.

Due to the overwhelming number of build breaks, I'm going to defer doing this to a later date, when a library-wide policy can be enacted. I'll update the PR to have `std::vector::size` call `__vector_layout::size`, etc. in a little bit, so that this PR isn't blamed for breaking ~everyone.

cjdb commented:
@ldionne @jyknight @rnk @philnik777 @mordante Ping. This PR has been open for 3.5 months now: would it be too much trouble to get some feedback in the next day or so, please?

jyknight commented on libcxx/include/__vector/vector.h:
```diff
@@ -82,25 +83,334 @@ _LIBCPP_PUSH_MACROS
+    return __begin_;
+  }
+
+  [[__nodiscard__]] _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI size_type __raw_sentinel() const _NOEXCEPT {
```
I find this use of the word "sentinel", as meaning "one of end or size") to be confusing. Usually I think of sentinel as "in-band data indicating you've reached the end". I don't have a better suggestion for a word to use here at the moment, though.

cjdb replied:
Two updates here:

1. I replaced the word `sentinel` with `boundary` (`bound` as of today) at the start of the year. I chose this word because the end pointer and the size both represent the boundary value for the range of valid values (either the raw end pointer, or `__begin_ + __size()`).
2. I've been able to make some serious improvements to how readable the respective layouts are. @ldionne helped me achieve this in two rather distinct ways. First, Louis helped me reduce how often `vector` directly accesses the layout's bound. Excluding type aliases, we're down now to four accesses. Louis also helped me calibrate how libc++ prioritises the DRY principle. I had oversubscribed on that, so his calibration helped me relax how much abstraction I was employing in the name of readability (which has improved readability more than the abstraction could have ever achieved).

Please take a look at the most recent implementation of `__vector_layout` and let me know what you think. My commit description for 0823c4205aabb4f77365633e92f090daa27c669d captures my goals with the abstract version, and why I've suddenly changed to what I'm proposing from here on.

jyknight commented on libcxx/include/__vector/vector.h:
```diff
@@ -82,25 +83,334 @@ _LIBCPP_PUSH_MACROS
+    return __alloc_;
+  }
+
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void swap(__vector_layout& __x)
```
This function can be written outside the ifdef, given `__raw_sentinel` and `__raw_capacity` accessors.

jyknight commented on libcxx/include/__vector/vector.h:
```diff
@@ -82,25 +83,334 @@ _LIBCPP_PUSH_MACROS
+    __size_ = static_cast<size_type>(__pos - __begin_);
+  }
+
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __set_capacity(size_type __cap) _NOEXCEPT { __cap_ = __cap; }
```
The same comment as for __set_sentinel above.

jyknight reviewed (commented):
I've taken a first pass over this, and have some design questions and comments.

Since I wasn't involved in previous discussions nor have I reviewed much libc++ code in the past, I apologize if my comments end up contradicting commonly accepted libc++ practices, or retreading already discussed issues.

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -82,25 +83,334 @@ _LIBCPP_PUSH_MACROS
 
 _LIBCPP_BEGIN_NAMESPACE_STD
 
+#ifdef _LIBCPP_ABI_SIZE_BASED_VECTOR
```
> Can you say why you created a separate `__vector_layout` type, vs ifdefs in the body of `class vector`? It feels like this factoring triggers the need for a LOT more modifications in this file than would otherwise be necessary. And I'm not sure what actual advantage is seen from having it.
>
> For example, none of the diffs relating to `__alloc_` and `__begin_` would be here, if not for the addition of the `__vector_layout` type.
>
> I think it certainly makes sense to group the ifdef-conditionalized code together by introducing internal helper functions -- as you have done. But that same effect can still be had without also introducing a new base class type.

Good question! The goal of `__vector_layout` was derived from `__split_buffer_(pointer|size)_layout`, which needs to be its own class because `vector` and `deque` require different layouts (there's currently no evidence that changing `deque`'s ABI to use `__split_buffer_pointer_layout` is beneficial, and it caused problems somewhere). We decided that it would be best from a software engineering standpoint to have a little redundancy in having a separate class, to keep `__split_buffer`'s logic as easily readable as possible (the trade-off was a more difficult PR diff, but substantially less context switching in the logic, and complete encapsulation). The changes that I made to `vector` rhyme very heavily with the changes made in #139632, so I figured it would be best to keep the design as similar as possible.

That said, I'm interested to explore your suggestion for two reasons:

1. You think it might reduce the diff, and if you're right, it'll probably help speed up reviewing this patch.
2. It might help with the debugger pretty-printers, something I'm still working on.

> On a trivial note, I'm not sure what the usual policy in libc++ is on ifdef ordering, but I was initially confused that the "old" layout was not textually the first ifdef'd half.

Interestingly, you're the first person to have brought this up (I expected it to have been flagged in the first patch). My rationale is that libc++ might want to support other layouts, and that ifdeffing is easier when done in a positive form, than a negative form.

rnk replied:
I was also going to question the need for a helper type, but I was coming at it from a debug info size perspective. If eliminating the `__compressed_pair` instantiation was valuable, eliminating `__vector_layout_type` might also be worth doing, if it's not providing valuable functionality or abstraction.

cjdb replied:
Done.

cjdb commented on libcxx/test/std/algorithms/alg.modifying.operations/alg.swap/ranges.swap_ranges.pass.cpp:
```diff
@@ -183,15 +167,34 @@ constexpr bool test() {
-    test_vector_bool<64>();
-    test_vector_bool<199>();
-    test_vector_bool<256>();
+  return true;
```
There were compile-time failures due to compiler limits on constexpr stack depths. We have two alternatives here:

1. Increase the limit Clang supports, which is a single-line change at the top of the file. Pro: single-line diff. Con: might need to increase the limit every so often.
2. Refactor the code so that it doesn't require that limit in the first place. Pro: won't run into this problem ever again. Con: this diff.

I could always split the refactoring into another PR, if it's too noisy for this one. That would also add documentation on why the change was necessary, which is probably a good enough reason for me to do it anyway.

cjdb replied:
Reverted in 832fdb07056eeb03eade88e5f712d4ed82511c81, and constexpr limit added in 47cffb725b4966364a2d8fb1c7b187a21d7e6118.

cjdb committed:
works `__vector_layout` into `vector`

[jyknight pointed out that we'd probably end up with a smaller diff][1]
(and simpler code) if we remove the `__vector_layout` type. [This also
reduces the volume of debuginfo, as pointed out by rnk][2].

[1]: https://github.com/llvm/llvm-project/pull/155330#discussion_r2607411335
[2]: https://github.com/llvm/llvm-project/pull/155330#discussion_r2612499610

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
further reduces the diff

0580dea significantly reduced the diff by returning the stable ABI code
to its original location. `__vector_base` did away with using `this->`
for most things, since it's largely redundant, but these were left out
in 0580dea. This commit adds most of those back.

cjdb committed:
fixes breaks from generic-cxx20

cjdb committed:
further reduces diff

cjdb committed:
gets GDB pretty-printer working

cjdb committed:
ensures that generic-cxx23 passes

cjdb commented on libcxx/utils/gdb/libcxx/printers.py:
```diff
@@ -351,17 +352,28 @@ def __init__(self, val):
             self.iterator = self._VectorBoolIterator(begin, self.length, bits_per_word)
         else:
-            end = self.val["__end_"]
+            for i in val.type.fields():
```
Fixed.

cjdb committed:
attempts to fix GCC issue

cjdb committed:
replaces alias to comply with GCC's -Wchanges-meaning

ldionne commented on libcxx/include/__vector/vector.h:
```diff
@@ -85,14 +86,18 @@ _LIBCPP_BEGIN_NAMESPACE_STD
+#else
+  template <class _SplitBuffer, class _Tp2, class _Allocator2>
+  using __split_buffer_layout _LIBCPP_NODEBUG = __split_buffer_pointer_layout<_SplitBuffer, _Tp2, _Allocator2>;
+#endif
```
In every place where you use `__split_buffer_layout`, you actually use `__split_buffer<_Tp, _Allocator, __split_buffer_layout>`. I think this would be simpler instead:

```c++
#ifdef _LIBCPP_ABI_SIZE_BASED_VECTOR
  using _SplitBuffer = __split_buffer<_Tp, _Allocator, __split_buffer_size_layout>;
#else
  using _SplitBuffer = __split_buffer<_Tp, _Allocator, __split_buffer_pointer_layout>;
#endif
```

cjdb replied:
I had this commit locally, but forgot to push on Friday afternoon. Should be up now.

ldionne commented on libcxx/include/__vector/vector.h:
```diff
@@ -577,25 +609,127 @@ class vector {
+    __cap_ = static_cast<size_type>(__pos - __begin_);
+  }
+
+  pointer __begin_  = nullptr;
```
Re https://github.com/llvm/llvm-project/pull/155330#discussion_r2607411335

> Can you say why you created a separate __vector_layout type, vs ifdefs in the body of class vector? It feels like this factoring triggers the need for a LOT more modifications in this file than would otherwise be necessary. And I'm not sure what actual advantage is seen from having it.

One of the main goals for introducing a distinct layout class was to have an explicit extension point with a clear interface that could be fulfilled in several different ways. For example, we could in principle introduce a third layout by just implementing a different layout class and changing one typedef inside `std::vector` itself.

We also wanted to get rid of as many `#ifdefs` as possible from the body of `std::vector`. Mixing `#ifdefs` with public interfaces is an invitation to get things wrong when we implement new features or make changes to existing public interfaces in response to WG21 papers or LWG issues, since we have more than one copy of the public interface to update. I am certain it's possible to find places in the library where we don't follow that guideline, but we definitely strive to.

As a result, I do have a marked preference for extracting two layout classes which are then used **with composition** from within `std::vector`.

cjdb replied:
I've compared 980deac7e0a20302b7d6b7c00f86bdb6427f8ee7 (PR HEAD~12) and bd4de75da409b19a199e1f2052a70e219b16bf3e (yours), and I think they're very similar, so I should be able to revert to 980deac7e0a20302b7d6b7c00f86bdb6427f8ee7 without much difficulty. I'm going to integrate as many of the differences from bd4de75da409b19a199e1f2052a70e219b16bf3e into my next revision (e.g. moving the layouts into `__vector/layout.h` was a nice suggestion!).

Pushing the next commit is blocked on a concern regarding one of the larger differences. Here's a really abstract version of bd4de75da409b19a199e1f2052a70e219b16bf3e:

```diff
 template<class _T, class _Alloc>
 class vector {
   // ...
-   pointer __begin_;
-   pointer __end_;
-   _LIBCPP_COMPRESSED_PAIR(pointer, __cap_ = nullptr, allocator_type, __alloc_);
+   __layout_t __layout_;
+   [[no_unique_address]] _Alloc __alloc_;
 };
```
My understanding is that this is ABI-breaking because we're replacing the compressed pair with something that's always `[[no_unique_address]]`. Is this understanding still correct, and if so, is it a break we're willing to carry forward? Most other incompatibilities are resolvable in a round of reviews, but this one seriously affects how much code I end up pushing.

cjdb replied:
I've just pushed a new layout type. What do you think about this particular design?

cjdb replied:
We've reached consensus on using a layout type. I'm resolving this comment in the interest of surfacing outstanding work items.

cjdb committed:
renames __split_buffer to comply with GCC's -Wchanges-meaning

cjdb committed:
Revert "replaces alias to comply with GCC's -Wchanges-meaning"

This reverts commit 81b360bf70d0dddb5e416159e43262f8f9c1655f.

ldionne reviewed (commented):
I spent this afternoon making myself familiar again with this PR and trying to see how some of my concerns could be addressed. At a high level, my main concerns are:

1. Extracting a separate layout class. I think this is important, that was actually one of the pre-requisites for upstreaming this change in the first place. That way, we're not just introducing new complexity to `std::vector` for the benefit of this ABI configuration. We're introducing an extension point which can be reused in the future for other experiments or other ABI configurations. I implemented the re-introduction of the layout class in https://github.com/llvm/llvm-project/commit/bd4de75da409b19a199e1f2052a70e219b16bf3e. Note that we should do it using composition, not inheritance like we did in `__split_buffer`, since that is conceptually simpler and opens the door to more simplifications by treating the layout as an independent object.
2. Figuring out a way to get rid of the notion of a sentinel which may be either a pointer or a size. That is very confusing when reading the code. I need to think about this more, but I suspect it might be possible to do it like I've sketched in one of the comments.

Other than that, the changes are pretty self contained and straightforward, so once we settle on the above, I think the rest should mostly sail through.

I've also extracted some minor stuff into #177022 and #177021 to minimize this diff since those were orthogonal changes.

I don't mind rebasing this patch myself once those land if you're OK with me pushing directly to your branch, since I know you've went through some churn already since this PR has been open.

cjdb committed:
[libcxx] fixes `__split_buffer_size_layout` bugs

**tl;dr**

As `__split_buffer` doesn't have any unit tests, and because #139632
disassociated adding `__split_buffer` support for both pointer-based
and size-based layouts from its `vector` counterpart, libc++ had no
way to expose `__split_buffer_size_layout` bugs until the size-based
vector patch integrated #139632. This commit fixes the two problems
that were identified while working on #155330.

**Context**

* `__split_buffer_size_layout` needs to update its size whenever
  `__begin_` is adjusted (relative to `__front_`).
* `__split_buffer`'s layout types originally updated `__begin_`
  independently of the sentinel type, via `__set_begin()` and
  `__set_sentinel()`, respectively.
* [Louis Dionne observed that `__set_begin()` was always coupled
  with `__set_sentinel()`][1], and suggested combining them into
  a single setter. This feedback was applied in ec20ccc.

**What went wrong (part 1)**

ec20ccc combined `__set_begin()` and `__set_sentinel()` into a single
function, but didn't remove the overlapping behaviour for
`__split_buffer_size_layout`. The code below shows how this caused a
problem:

```
void __set_valid_range(pointer __new_begin, pointer __new_end)
{
  __size_ -= __new_begin - __begin_;
  __begin_ = __new_begin;
  __set_sentinel(__new_end);

  // Inlining call to `__set_sentinel(__new_end)`:
  // __size_ += __new_end - end();
  //
  // Inlining call to `end()`:
  // __size_ += __new_end - (__begin_ + __size_);
}
```

`__set_valid_range(pointer, pointer)` sets `__size_` on both sides of
setting `__begin_`. While either is valid on its own, applying both is
incorrect. `__set_valid_range(pointer, size_type)` has the same abstract
logic, but only wastes compute cycles, rather than affecting
correctness:

```
void __set_valid_range(pointer __new_begin, size_type __new_size)
{
  __size_ -= __new_begin_ - __begin_;
  __begin_ = __new_begin_;
  __set_sentinel(__new_size);

  // Inlining call to `__set_sentinel(__new_size)`:
  // __size = __new_size;
  //
  // First line is an efficiency bug, not a correctness bug.
}
```

This commit removes the first line both overloads. Additionally, this
commit optimises `__set_sentinel(pointer)` computes, as described
below:

```
__set_sentinel(__new_end)
  ≡ __size_ += __new_end - end()
  ≡ __size_ += __new_end - (__begin_ + __size_)
  ≡ __size_  = __size_ + __new_end - (__begin_ + __size_)
             = __size_ + __new_end - __begin_ - __size_
             = __new_end - __begin_
```

**What went wrong (part 2)**

The second bug was a type error in `__swap_without_allocator`. As most
function signatures identical in both layout types, they were copied
from `__split_buffer_pointer_layout` into `__split_buffer_size_layout`
to ensure that they were correct. `__swap_without_allocator` is a member
function that operates on the layout type directly, and so its signature
needs to differ with each layout type. Although other member functions
were adjusted (e.g. `swap`), `__swap_without_allocator` was not, leading
to compile-time errors when working on size-based vector, since
`__split_buffer_pointer_layout` is not used when `vector` is size-based.

**How this ended up in the main branch**

Since the size-based vector project applies very similar changes to
`__split_buffer` and `std::vector`, it was split into two PRs, so that
reviewers didn't need to provide duplicate feedback. The PRs were split
on types: #139632 patches `__split_buffer` and #155330 patches `vector`.
Since `vector` needs a layout-compatible `__split_buffer`, #139632 was
merged before #155330, to keep the PR story as simple as possible.

`__split_buffer` is an implementation detail and does not have its own
unit tests: correctness bugs are being surfaced by its drivers, namely
`std::vector` and `std::deque`. As such, there isn't a way to identify
that `__split_buffer_size_layout` was correct without paradoxically
testing it alongside a size-based vector implementation.

**How this can be avoided in the future**

There are two primary ways to avoid this type of problem moving forward.

1. **Split work across logic instead of types.** #139632 could have
   introduced the pointer layouts for both `__split_buffer` and
   `vector`, which would have set the structure for both types, but not
   made any material changes to the implementation. This would have
   coupled dependencies instead of typenames, and possibly simplified
   the review process further, as no logic would have changed.

   The trade-off for this approach is that both layouts influenced the
   layout's API design, and might have initially favoured the pointer
   layout, leading to adjustments later on.
2. **Stack dependent PRs, and submit when all have been LGTM'd.** This
   approach would allow the PRs to be sent for review as they were, but
   ensures that logic that's coupled across commits is checked in
   together. Stacking the PRs would allow CI to pick up both PRs, which
   is critical for ensuring that the dependent PR is correct.

   This was considered for #155330, but the GitHub model for [stacking
   PRs][2] required some work to be done ahead of #139632's creation
   without the risk of confusing reviewers. Closing #139632 and opening a
   new PR in service wasn't ideal, as it would have separated the PR
   history from the PR that was ultimately merged.

**How this patch was tested**

Since this commit is a change to `__split_buffer_size_layout`, and needs
to be merged with `main` before #155330, this commit was cherry-picked
into #155330, and tested using the libc++ buildbot job for the unstable
ABI.

[1]: https://github.com/llvm/llvm-project/pull/139632/#discussion_r2282535795
[2]: https://llvm.org/docs/GitHub.html#stacked-pull-requests

cjdb committed:
reverts ranges.swap_ranges.pass.cpp

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
explicitly casts `__other` to `__base_type&`

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
reintroduces `__vector_layout`

This makes it much closer in design to ba94eeb. Further effort has gone
into reducing the diff with ToT.

cjdb committed:
limits how often vector needs to manually set `__vector_layout`

cjdb committed:
teaches GDB how to use the new vector layout

cjdb committed:
adds missing compressed_pair.h

cjdb committed:
runs clang-format

cjdb commented on libcxx/include/__split_buffer:
```diff
@@ -188,6 +188,13 @@ public:
     __back_cap_  = __other.__back_cap_;
   }
 
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void
```
The changes in this file can be moved to a new PR if necessary.

cjdb committed:
fixes formatting issues flagged by CI

cjdb committed:
fixes C++03 and relevant GCC failures

cjdb committed:
works around a Clang 20 bug

cjdb committed:
formats and removes pretty Unicode characters

cjdb committed:
adds some non-const overloads

cjdb committed:
[libcxx] tightens `__add_alignment_assumption` return type

`__add_alignment_assumption` unilaterally returned a `pointer`, even
when passed a `const_pointer`. This was surfaced by some (but not all)
CI jobs when rebasing `std::vector` to have a layout type.

cjdb committed:
adds missing __capacity_ptr

ldionne commented on libcxx/include/__vector/vector.h:
```diff
@@ -409,50 +424,50 @@ class vector {
   [[__nodiscard__]] _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI reference back() _NOEXCEPT {
     _LIBCPP_ASSERT_VALID_ELEMENT_ACCESS(!empty(), "back() called on an empty vector");
-    return *(this->__end_ - 1);
+    return __base_type::back();
```
I would suggest naming the layout's method `__back()` to make it clear that it's an internal API, not a public API. This is something we've done in other places.

ldionne commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,443 @@
+    return __alloc_;
+  }
+
+  /// Returns a reference to the stored allocator.
```
```suggestion
```

Since it's the same as the non-const version above.

ldionne commented on libcxx/include/__vector/vector.h:
```diff
@@ -862,47 +857,52 @@ template <ranges::input_range _Range,
+      std::__to_address(__new_begin));
   __v.__set_valid_range(__new_begin, __v.end());
-  __end_ = __begin_; // All the objects have been destroyed by relocating them.
+  __set_boundary(static_cast<size_type>(0)); // All the objects have been destroyed by relocating them.
```
I just realized that something quite subtle is happening here. I presume this would be an ambiguous call if you dropped `static_cast<size_type>` since it could either be trying to call `__set_boundary(pointer)` or `__set_boundary(size_type)`, is that right?

cjdb replied:
Yes. I think the right thing to do here is to use `__vector_layout::__zero_boundary_type()`, which not only gets rid of the subtle `static_cast`, but also lets us call `__set_boundary(pointer)` for the pointer-based layout.

ldionne commented on libcxx/include/__vector/vector.h:
```diff
@@ -862,47 +857,52 @@ template <ranges::input_range _Range,
-  __v.__set_valid_range(__new_begin, __v.end());
-  __end_ = __begin_; // All the objects have been destroyed by relocating them.
+      this->__alloc(), std::__to_address(__begin_ptr()), std::__to_address(__p), std::__to_address(__new_begin));
+  __v.__set_valid_range(__new_begin, __v.size() + size());
```
Similarly, here, why doesn't this translate directly to `__v.__set_valid_range(__new_begin, __v.end())`?

cjdb replied:
See [one comment below](https://github.com/llvm/llvm-project/pull/155330#discussion_r2774456260).

ldionne commented on libcxx/include/__vector/vector.h:
```diff
@@ -1247,17 +1247,18 @@ template <class _Tp, class _Allocator>
     }
   } else {
-    __split_buffer<value_type, allocator_type> __v(__recommend(size() + 1), __p - this->__begin_, this->__alloc_);
+    _SplitBuffer __v(__recommend(size() + 1), __p - this->__begin_ptr(), this->__alloc());
```
One suggestion I just thought about: we could refactor the code *prior* to this PR to use `this->__begin_ptr()`, `this->__end_ptr()` and `__this->__alloc()`, that would be a NFC change but it would reduce the size of this diff. That would be effectively a revert of #114423, but oh well, needs change and so does code.

I'm not requesting that you make that refactoring, but here as well, if you're up for it I think this would reduce the size of this diff and perhaps allow us to focus on the more important parts of this patch with less distraction. It might also make your rebasing easier.

cjdb replied:
I'm not sure about this one. Your suggestion works well for the current PR and I'd take you up on it if we were sticking with that code model. Changing to composition means that `this->` becomes `__layout_.`, so we're still taking a diff here. Am I missing something? I'll happily make the PR if it reduces the diff.

ldionne commented on libcxx/include/__vector/vector.h:
```diff
@@ -1267,25 +1268,26 @@ vector<_Tp, _Allocator>::emplace(const_iterator __position, _Args&&... __args) {
       }
     } else {
-      __split_buffer<value_type, allocator_type> __v(__recommend(size() + __n), __p - this->__begin_, this->__alloc_);
+      _SplitBuffer __v(__recommend(size() + __n), __p - this->__begin_ptr(), this->__alloc());
```
I guess we could also extract the trivial renaming of `_SplitBuffer` into its own PR: that would make the diff even smaller and that patch would also be trivial to approve.

ldionne reviewed (commented):
Thanks for all the work on this PR! I think this is really going into the right direction.

I have a few suggestions for extracting stuff out of this PR that would allow us to focus on the last remaining "points of contention", the most important of which is IMO about expressions whose type may be either a size or a pointer, leading to difficult to understand code (I left comments explaining what I mean).

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -84,19 +86,33 @@ _LIBCPP_PUSH_MACROS
-class vector {
-  template <class _Up, class _Alloc>
-  using __split_buffer _LIBCPP_NODEBUG = std::__split_buffer<_Up, _Alloc, __split_buffer_pointer_layout>;
+class vector : __vector_layout<_Tp, _Allocator> {
```
I'm partial to this request. @philnik777 asked me to [use inheritance for the `__split_buffer` layouts](https://github.com/llvm/llvm-project/pull/139632#discussion_r2259783216), so I was aiming to keep the two models consistent.

That said, I think I'm going to need to refactor `__split_buffer` so that its layout types are consistent with the improved `__vector_layout`, so changing inheritance to composition can be a part of the refactoring over there.

cjdb committed:
adjusts documentation

cjdb committed:
reverts default constructor

cjdb commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,443 @@
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __set_boundary(pointer __end) _NOEXCEPT;
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __set_capacity(size_type __size) _NOEXCEPT;
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __set_capacity(pointer __end) _NOEXCEPT;
+  // Works around a Clang 20 zero-initialization bug on user-defined pointer types.
```
The CI for ebbe89432ae9ee2de82f1660bd936a7d7d7a492a (plus local debugging) shows that Clang 20 has a problem with using `{}` to create a zero-initialised object. Based on my understanding of zero-initialisation and the fact that it's only apparent on Clang 20, I concluded that it's a bug, but I should probably confirm that.

@AaronBallman can you confirm whether it's a bug, and if so, do you know which PR resolved it?

AaronBallman replied:
I can't seem to confirm it, but maybe I'm holding your test case incorrectly: https://godbolt.org/z/dPx3TTTab

cjdb replied:
Thanks for suggesting this, it's a much more elegant solution!

cjdb committed:
changes from inheritance to compositon

See https://github.com/llvm/llvm-project/pull/155330/changes#r2774234713.

cjdb committed:
adds assertions

cjdb committed:
uglifies __vector_layout members

cjdb committed:
renames parameters to improve readability

cjdb committed:
adds `__zero_relative_to_begin` and reset methods to `__vector_layout`

Using `static_cast<size_type>(0)` was a fairly subtle way to avoid
overload resolution ambiguity between `size_type{0}` and `nullptr`.
This commit intends to improve readability by replacing that cast
with a named function to communicate intent.

This commit introduces a method that resets `__begin_`, `__boundary_`,
and `__capacity_` together. Since the capacity is only ever zeroed out
when `__begin_` and `__boundary_` are also zeroed out, so these should
be executed by a single function call.

cjdb committed:
moves `__swap_out_circular_buffer` to `__vector_layout`

cjdb committed:
moves `__relocate` to `__split_buffer`

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -862,47 +857,52 @@ template <ranges::input_range _Range,
-  __end_ = __begin_; // All the objects have been destroyed by relocating them.
+      this->__alloc(), std::__to_address(__begin_ptr()), std::__to_address(__p), std::__to_address(__new_begin));
+  __v.__set_valid_range(__new_begin, __v.size() + size());
+  __set_boundary(static_cast<size_type>(0)); // All the objects have been destroyed by relocating them.
```
> You need to have a single overload of `__set_boundary` that takes both possible expression types, however I think that's a bug, not a feature. So I think that making a change like `__set_end` and `__set_size` would force us to spot places where expressions have a "undecided" type, and figure out a proper way to make those work.

The reason we need an overload set is because I'm trying to combine the size-based vector project's goal of eliminating as much pointer arithmetic as possible with the broader libc++ goal of maximum code reuse. Erasing the type as `__boundary_type` helps to eliminate excess pointer arithmetic because we don't need to do pointer-to-size or size-to-pointer conversions when setting the bound. That said, I've managed to get `vector` into a position where only one call to `__set_boundary` is dependent on `__boundary_type` (`__vector_layout` can't take advantage of this without code duplication).

We can absolutely bifurcate this sort of stuff, but I'm concerned that doing so will ultimately lead to either a situation where it would be easier to maintain as two completely independent `vector` definitions, or one where the size-based `__vector_layout` isn't as competitive as the one that I've benchmarked.

> Note that I'd be willing to go as far as refactoring `__swap_out_circular_buffer` into something different, perhaps even part of the layout classes, if that's what we need to eliminate this kind of macro-like-just-not-with-the-preprocessor code.

I think this is the best path forward. I've put it into `__split_buffer`, and will factor that out into its own patch if you like how I've done things.

ldionne replied:
> The reason we need an overload set is because I'm trying to combine the size-based vector project's goal of eliminating as much pointer arithmetic as possible with the broader libc++ goal of maximum code reuse.

To clarify, I don't think we necessarily maximum code reuse mechanically speaking. I think what's more important is to minimize the duplication of code when there's actual logic involved, and especially minimizing the amount of "branches" within that logic (but we're good on that with the split to a separate vector layout type).

**Clarifying edit:** So, in a way, I'd be fine with more duplication if the duplicated code is only e.g. logic-less declarations in the vector layout interface, or similar.

> Erasing the type as `__boundary_type` helps to eliminate excess pointer arithmetic because we don't need to do pointer-to-size or size-to-pointer conversions when setting the bound. That said, I've managed to get vector into a position where only one call to `__set_boundary` is dependent on `__boundary_type` (`__vector_layout` can't take advantage of this without code duplication).

Do you have a good idea of what it would entail to change the code such that we have different `__set_size` and `__set_end` functions? I think it would only impact `vector<_Tp, _Allocator>::emplace_back(_Args&&... __args)`, is that right?

cjdb committed:
replaces __set_valid_range with __set_layout

cjdb committed:
clang-format fix

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
removes unnecessary const-qualifier

cjdb committed:
Merge remote-tracking branch 'refs/remotes/cjdb/size-based-vector-part2' into size-based-vector-part2

ldionne commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,443 @@
+///    and boxes with `xx` have been allocated, but aren't being used as elements right now.
+//
+/// We conducted an extensive A/B test on production software to confirm that the size-based layout
+/// improves compute performance by 0.5%, and decreases system memory usage by up to 0.33%.
```
I think we should drop this paragraph entirely:

> We conducted an extensive A/B test on production software to confirm that the size-based layout
> improves compute performance by 0.5%, and decreases system memory usage by up to 0.33%.

This belongs in the PR description, in any RFC, in a talk, etc, but not in the documentation. We already do say the important part above:

> Programs using the size-based layout have been measured to yield improved compute and memory performance over the pointer-based layout.

We also recommend people to use this new layout in the absence of ABI constraints. IMO, going beyond that bypasses the purpose of in-code documentation.

cjdb committed:
fixes who accesses `__capacity_ptr()`

ldionne commented on libcxx/include/__split_buffer:
```diff
@@ -188,6 +189,57 @@ public:
+    std::swap(__back_cap_, __back_capacity);
+  }
+
+  /// `__relocate` relocates the objects in `[__begin_, __size())` into the front of `__buffer` and
```
```suggestion
  /// `__relocate` relocates the objects in `[begin(), end())` into the front of `__buffer` and
```

Since this is about semantics, I think this is the clearest way to describe those semantics.

Also, what is `__buffer`? It's not a parameter to the function.

ldionne reviewed (commented):
Thanks a lot for the discussions and your changes so far. This is shaping up! This is a much, much more difficult review than I initially expected when this project was discussed, but we'll get there 😅

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
fixes botched merge

cjdb committed:
documentation changes

cjdb committed:
[libc++] Add `__exchange` as a C++11 utility

`std::exchange` is helpful for simplifying move operations because we
can couple the resetting component with the move operation. This both
halves the number of necessary lines for each move, and also ensures we
haven't forgotten to reset the moved-from object.

`std::exchange` is a C++14 feature, so we'll need to use `__exchange`
for anything that's from C++11.

cjdb reviewed (commented):
@ldionne Thanks for reviewing! I think I've addressed (but not necessarily fixed) all comments.

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
fixes merge issues

ldionne reviewed (commented):
Main remaining items:
- Remove `__zero_boundary_type` and `__zero_from_begin()`
- Don't zero-out moved-from layout in the layout move ctor
- Split `__set_boundary` into two functions and apply the changes that cascade from that (a few renames and the added `#ifdef` in `emplace_back`).

I think this can go in once these have been addressed!

cjdb committed:
updates docs

cjdb committed:
renames `__vector_layout::__move_without_allocator`

cjdb committed:
simplifies `vector::emplace_back`

cjdb committed:
removes headers

cjdb committed:
fixup for emplace_back

cjdb committed:
renames setters

cjdb committed:
removes zeroers

cjdb committed:
bifrucates `__vector_layout` implementations to improve readability

The original upstreaming of size-based vector used the overload set
`__set_boundary` to set the size/end pointer in order to minimise
code duplication. The two overloads were `__set_boundary(size_type)`
and `__set_boundary(pointer)`. The patch preferred an overload set
over `__set_size` and `__set_end_pointer` for two reasons:

1. Calls typically occurred in layout-agnostic sections of code.
   Arguments would sometimes match the layout's representation, and
   hard-coding `__set_end_pointer` when using a size-based vector that
   wants to set a size would require converting the size to an end
   pointer, and then back to a size. Alternatively, it would force
   layout-agnostic code to favour a specific layout, which was
   undesirable.
2. Names such as `__set_size` and `__set_end_pointer` are highly
   layout-specific. A size-based layout doesn't have an end-pointer, so
   using `__set_end_pointer` is a readability issue. The workaround for
   this is to manually perform overload resolution on `__set_boundary`:
   `__set_size` becomes `__set_boundary_using_integer` and
   `__set_end_pointer` becomes `__set_boundary_using_pointer`.

Several reviewers have expressed concerns that `__set_boundary` is a
readability concern. The review process also consolidated most calls
to `__set_boundary` from `vector` into `__vector_layout`. d49927e
renamed the setters to have individual names, but this ran into the
problems outlined above. This means the setters no longer serve their
original purpose, and hurt readability.

The original code was written to absolutely minimise code duplication.
Since libc++ is okay with repetition that improves readability, this
commit refactors `__vector_layout` to define any function that refers
to a size or end pointer as `__size_` and `__end_`, respectively.
Specifically-named members substantially improve readability over using
any type of DRY abstraction. Since readability is the primary goal for
teasing out more explicit names for `__set_boundary`, it seems like this
is an appropriate change to apply.

cjdb committed:
removes accidental definition

cjdb committed:
renames `swap` to `__swap`

The `swap` function turned out to be a member function, so we can rename
it to `__swap` without being concerned about the ADL idiom.

cjdb committed:
removes experimental `std::exchange`

cjdb committed:
changes how `__capacity_` is initialised

cjdb committed:
runs `git clang-format`

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -1129,18 +1087,18 @@ _LIBCPP_CONSTEXPR_SINCE_CXX20 inline
+      [&] { __current_boundary = __emplace_back_slow_path(std::forward<_Args>(__args)...); });
 
-  this->__end_ = __end;
+  __layout_.__set_boundary(__current_boundary);
```
It turns out that `__emplace_back_assume_capacity` and `__emplace_back_slow_path` update the layout, so `emplace_back` doesn't need to do that at all. That means we don't need to bifurcate the `emplace_back` logic!

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -740,14 +757,14 @@ class vector {
 
     _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI ~_ConstructTransaction() {
-      __v_.__end_ = __pos_;
+      __v_.__layout_.__set_boundary(__pos_);
```
I find the name `__set_end` to be misleading whenever a size-based vector is being used because we're not setting the end: we're doing arithmetic to set the size. It's not a problem anymore, but I hold similar opinions towards `__set_size`.

Do you have any objections to `__set_bound_using_pointer`? It's a bit of a mouthful, but it makes it very clear that we're (a) setting some abstract object that represents the vector's bound (what I want), and (b) explicitly describes what we're using to set that bound (what you and @jyknight want).

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -760,10 +777,10 @@ class vector {
-      __alloc_traits::destroy(this->__alloc_, std::__to_address(--__soon_to_be_end));
-    this->__end_ = __new_last;
+      __alloc_traits::destroy(this->__layout_.__alloc(), std::__to_address(--__soon_to_be_end));
+    __layout_.__set_boundary(__new_last);
```
Resolving as a duplicate issue of the above to declutter the list of open comments. This needs to be updated whenever the previous one is anyway.

cjdb commented on libcxx/include/__split_buffer:
```diff
@@ -316,9 +314,9 @@ public:
-      __split_buffer_pointer_layout<__split_buffer<value_type, __alloc_rr&, __split_buffer_pointer_layout>,
-                                    value_type,
-                                    __alloc_rr&>& __other) _NOEXCEPT {
+      __split_buffer_size_layout<__split_buffer<value_type, __alloc_rr&, __split_buffer_size_layout>,
```
We ended up merging this in #178341, so I'm going to resolve this comment.

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -82,25 +83,334 @@ _LIBCPP_PUSH_MACROS
+    return __cap_ - __size_;
+  }
+
+  [[__nodiscard__]] _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI bool __is_full() const _NOEXCEPT {
```
Resolving because this comment is redundant: please see https://github.com/llvm/llvm-project/pull/155330/changes/ba94eeba1e22ebee9708c770c3b7da73eaad131e#r3163658583 for why.

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -82,25 +83,334 @@ _LIBCPP_PUSH_MACROS
+    __size_  = __size;
+  }
+
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __set_sentinel(size_type __size) _NOEXCEPT {
```
Resolving because this comment is redundant: please see https://github.com/llvm/llvm-project/pull/155330/changes/ba94eeba1e22ebee9708c770c3b7da73eaad131e#r3163658583 for why.

cjdb commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,508 @@
+        __alloc_(std::move(__other.__alloc_)) {
+    __other.__begin_    = nullptr;
+    __other.__boundary_ = __zero_boundary_type();
+    __other.__capacity_ = __zero_boundary_type();
```
I've done this, but I still think this should be handled by the layout's move constructor.

Although `vector` owns the resources, `__vector_layout` still owns its own members. We typically consider it to be healthy for user-defined move constructors to also reset the values that they're moving. It's also worth noting that `vector` is explicitly choosing to reset the layout by using `std::move` instead of copying. Finally, and this might be the most important reason: explicitly resetting is error-prone: we need to be careful to do `__x.__layout_.__reset()`, not `__layout_.__reset()`.

I'll go ahead with whichever direction you feel is best.

cjdb commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,443 @@
+  size_type __capacity_ = 0;
+  [[no_unique_address]] allocator_type __alloc_;
+#else
+  pointer __boundary_ = nullptr;
```
I've taken you up on this. I said this elsewhere, but you helped me calibrate libc++'s preferences for DRY, so I felt empowered to go ahead with this change.

I'm going to resolve this comment (in particular), so we don't accidentally end up discussing the same thing in multiple places. Let's make https://github.com/llvm/llvm-project/pull/155330/changes/ba94eeba1e22ebee9708c770c3b7da73eaad131e#r3163658583 the canonical discussion point, so that James is also included.

cjdb commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,505 @@
+
+template <class _Tp, class _Alloc>
+_LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI typename __vector_layout<_Tp, _Alloc>::__boundary_type
+__vector_layout<_Tp, _Alloc>::__zero_boundary_type() _NOEXCEPT {
```
Removed in favour of your suggestion.

cjdb commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,443 @@
+    __other.__begin_    = nullptr;
+    __other.__boundary_ = __zero_boundary_type();
+    __other.__capacity_ = __zero_boundary_type();
+  }
```
Resolving since your more recent comment about renaming this to `__move_assign_without_allocator` seems to indicate we've reached consensus on this.

cjdb commented on libcxx/include/__vector/vector.h:
```diff
@@ -955,22 +922,15 @@ _LIBCPP_CONSTEXPR_SINCE_CXX20 inline _LIBCPP_HIDE_FROM_ABI vector<_Tp, _Allocato
-    __x.__begin_ = __x.__end_ = __x.__cap_ = nullptr;
+    : __layout_(__a) {
+  if (__a == __x.__layout_.__alloc()) {
+    __layout_.__move_without_allocator(__x.__layout_);
```
I like this name too. Done.

cjdb committed:
changes who resets a moved-from layout

cjdb reviewed (commented):
Thank you for your review. I'm really happy that we're starting to reach consensus on this! I think I've addressed as much as possible without further input.

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
fixes merge

cjdb committed:
changes to __exchange to help debug macOS issues

cjdb committed:
Teach LLDB's pretty printers to read the new vector layout

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
runs clang-format

cjdb committed:
Update LLDB pretty-printer

cjdb committed:
Refactor libc++'s LLDB `std::vector` pretty-printer

cjdb committed:
Fix CI issues

cjdb committed:
formats patch and updates test config

cjdb committed:
removes macros from previous implementation

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
Incorporate recent review feedback

cjdb committed:
document size-based vector in ABIGuarantees.rst

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb commented on libcxx/include/__configuration/abi.h:
```diff
@@ -77,6 +77,7 @@
 #  define _LIBCPP_ABI_NO_REVERSE_ITERATOR_SECOND_MEMBER
 #  define _LIBCPP_ABI_OPTIMIZED_FUNCTION
 #  define _LIBCPP_ABI_REGEX_CONSTANTS_NONZERO
+#  define _LIBCPP_ABI_SIZE_BASED_VECTOR
```
Done.

ldionne commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,545 @@
+template <class _Tp, class _Alloc>
+_LIBCPP_CONSTEXPR_SINCE_CXX20 typename __vector_layout<_Tp, _Alloc>::pointer
+__vector_layout<_Tp, _Alloc>::__capacity_ptr() _NOEXCEPT {
+  return __begin_ + __size_;
```
```suggestion
  return __begin_ + __capacity_;
```

cjdb replied:
Done.

ldionne commented on libcxx/docs/ABIGuarantees.rst:
```diff
@@ -58,6 +58,54 @@ most cases.
 This significantly restructures how ``function`` is written to provide better performance, but is currently not ABI
 stable.
 
+``_LIBCPP_ABI_VECTOR_LAYOUT``
```
I like that we're documenting the ABI macro here, I think that was an oversight in previous versions of the patch. However, the name of the macro is `_LIBCPP_ABI_SIZE_BASED_VECTOR`, and `_LIBCPP_ABI_VECTOR_LAYOUT_POINTER_BASED` / `_LIBCPP_ABI_VECTOR_LAYOUT_SIZE_BASED` don't exist. Instead, I think we should follow the style used to document other macros a bit more closely. I'd suggest:

```
``_LIBCPP_ABI_SIZE_BASED_VECTOR``
-----------------------------------------
This changes the layout of `std::vector` from <...> to <...>.
```

You can keep the meat of your current documentation in the `<...>` in a form that makes sense.

cjdb replied:
Done.

ldionne commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,545 @@
+    return __size_ == 0 && __capacity_ == 0;
+  return __size_ <= __capacity_;
+}
+#else
```
```suggestion
#else // pointer-based layout
```

Since this is a large block, it helps navigation.

cjdb replied:
Done.

ldionne commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,545 @@
+/// `std::vector` can be represented in a variety of ways. Each representation strongly influences
+/// the codegen when calling vector operations, which can significantly impact runtime performance
+/// and memory utilisation. libc++ provides two alternative layouts for `std::vector`, although only
+/// one can be active for an enitre binary:
```
```suggestion
/// one can be active for an entire binary:
```

cjdb replied:
Done.

ldionne commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,545 @@
+
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __reset_without_allocator() _NOEXCEPT;
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __swap(__vector_layout& __other) _NOEXCEPT;
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __swap_layouts(_SplitBuffer& __other) _NOEXCEPT;
```
I think this is dead code (and also the two definitions).

cjdb replied:
Removed.

ldionne commented on libcxx/include/__split_buffer:
```diff
@@ -196,6 +197,57 @@ public:
+  ///   * `__front_spare() == __pivot - __first`
+  ///   * `__back_spare() == __last - __pivot`
+  /// Exceptions: This function has a strong exception guarantee if `__first == __pivot` or
+  /// `__last - __first == __pivot`.
```
What do you mean here? `__last - __first` is a size, but `__pivot` is a pointer, so this doesn't type check. Did you mean `__last == __pivot`?

cjdb replied:
Yeah, it looks like that's correct. I think a previous iteration might've been `__last - __first == size()`, and I didn't properly update the docs.

ldionne commented on libcxx/include/__vector/vector.h:
```diff
@@ -462,19 +468,24 @@ class vector {
 
   template <class... _Args>
-  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI __emplace_back_result_t emplace_back(_Args&&... __args) {
```
This function was previously written using `__if_likely_else` on purpose (to workaround something really subtle with Clang), see the discussion in #94379. Unless the underlying Clang issue has been fixed, this refactoring would unfortunately introduce a codegen regression. Since this is a very important function for `vector` (arguably the most important one?), I think we should go back to using it.

Now, regarding whether we can avoid setting `__end = __emplace_back_slow_path(std::forward<_Args>(__args)...);` from this function, I think we'd want to check the codegen before and after. IIRC, the idea was to make the flow of data explicitly visible to the compiler because it allowed it to generate better code (that's why we didn't call `back()` again in the pre-refactor version).

We're on the same page that the new code is better in the sense of readability and reuse, but we had decided to make a conscious readability tradeoff since this function is very performance sensitive.

Now, if we find out that we can avoid the `_end = __emplace_back_slow_path(...)` without hurting code generation (I'd be surprised because something would have had to change), that means we can avoid bifurcating `__emplace_back`. Otherwise, I'd bifurcate it for the size/pointer based vector layouts and accept it as tech debt.

cjdb replied:
Let's defer this change to a later PR (that I'll open ~soon). I had some fairly promising [codegen experiments](https://carbon.godbolt.org/z/PTqTrWs7d) with removing `__if_likely_else` before pushing the commit for review. I've been trying to get some macrobenchmarks since you've flagged, but it looks like there's a few blockers for a timely turnaround on that, and I don't want to block on something that can be done independently.

cjdb replied:
> Now, regarding whether we can avoid setting `__end = __emplace_back_slow_path(std::forward<_Args>(__args)...);` from this function, I think we'd want to check the codegen before and after. IIRC, the idea was to make the flow of data explicitly visible to the compiler because it allowed it to generate better code (that's why we didn't call `back()` again in the pre-refactor version).
> 
> We're on the same page that the new code is better in the sense of readability and reuse, but we had decided to make a conscious readability tradeoff since this function is very performance sensitive.
> 
> Now, if we find out that we can avoid the `_end = __emplace_back_slow_path(...)` without hurting code generation (I'd be surprised because something would have had to change), that means we can avoid bifurcating `__emplace_back`. Otherwise, I'd bifurcate it for the size/pointer based vector layouts and accept it as tech debt.

I modified the above experiment to look at [what happens when I toggle this change](https://carbon.godbolt.org/z/WMG1qT5G3) (you'll want to look for `IMPLICIT_SET_END`). I'm by no means an expert in assembly, but it looks like we get simpler codegen in both implementations across all experiments to me, when we stop assigning `__end` in `emplace_back`?

cjdb replied:
I've rolled this back for the time being, and will make a follow-up PR once the benchmark results have come through.

ldionne replied:
Thanks. The original change that introduced this pattern explained the reason behind it: 6fe4e033f07d332980e1997c19fe705cff9d07a4

Doing [my own test similar to yours](https://godbolt.org/z/Ta6ov4d6K), I also don't see an immediate problem with dropping the explicit store. But I'd rather do it intentionally in a dedicated PR where we can fully investigate.

ldionne commented on libcxx/include/__vector/layout.h:
```diff
@@ -0,0 +1,545 @@
+  /// outside `__vector_layout`. Its result must be used with type deduction to avoid compile-time
+  /// failures.
+  [[__nodiscard__]] _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI __bound_type
+  __bound_representation() const _NOEXCEPT;
```
This can be removed now since it doesn't need to be returned from `__emplace_back_slow_path`. Also, `__bound_type` can be removed.

Note that this is true regardless of what we end up doing about `emplace_back` (see my other comment), since we can unconditionally return a pointer from `__emplace_back_slow_path` if we need that.

cjdb replied:
Done.

cjdb committed:
Add missing `typename` specifier to satisfy GCC in C++11 mode

cjdb committed:
Run clang-format

cjdb committed:
Revert __emplace_back_slow_path optimisation for now

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
Fix formatting in `ABIGuarantees.rst`

cjdb committed:
Merge branch 'main' into size-based-vector-part2

cjdb committed:
Add missing `_LIBCPP_NODEBUG` to detail type

ldionne reviewed (approved):
This has been a long time coming, but I think this is good to go. Thanks for the diligence and patience on this review, this was non-trivial.

Since this touches `vector` in non-trivial ways, I do expect we might need some follow-ups on this. I don't know exactly what, but I wouldn't be surprised if something came up, since it often does when we touch something as ubiquitous (e.g. performance change, debug size impact, or else). Let's just keep an eye out for this, and when/if this happens, we'll decide whether to temporarily roll-back or fix forward.

Thanks!
